### PR TITLE
New Loadflow parameters

### LINKED ISSUE #1293: Loadflow parameters and result v2
* **Do you want to request a *feature* or report a *bug*?**
feature

* **What is the current behavior?**
`LoadFlowParameters` is really poor in the current version. It only have a small subset of load flow features and as a consequence all loadflow implementation rely on an extension to provide additional features. And those extensions have lots of common features...

Same for LoadFlowResult which rely on a very weak metrics Map<String, String> as a main output informations.

* **What is the expected behavior?**
Most of the parameters should move to `LoadFlowParameters` and not all implementation need to support all options. If a feature is asked but not supported by a particular impl an exception or a warning could be raised by the implementation

Based on Open LoadFlow and Hades2 here is a list of additional feature to add to LoadFlowParameters:

  - distributed slack (boolean)
  - slack bus id (in case of manual slack bus selection) 
  - balance type (PROPORTIONAL_TO_GENERATION_P, PROPORTIONAL_TO_GENERATION_P_MAX, PROPORTIONAL_TO_LOAD)
  - DC mode (boolean)
  - generator voltage remote control (boolean)
  - countries to balance (set<Country>)
  - transformer voltage control (boolean)
  - max iterations (int)
  - min and max realistic voltage (2 floats)
  - minimum reactance (SI to be compatible with different PU)
  - compute all connected components
  - ???

And to LoadFlowResult:
  - selected slack bus
  - slack bus injection
  - ???

* **What is the motivation / use case for changing the behavior?**
To facilitate load flow parameters creation.

* **Please tell us about your environment:**
  - PowSyBl Version: ...
  - OS Version: ...

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, spectrum, etc)

(if a question doesn't apply, you can delete it)

### PR DESCRIPTION
**Pull request by tadam50**
Signed-off-by: Thomas ADAM <tadam@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
fix #1293

### CODE REVIEW

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 1]
```diff
@@ -58,6 +59,8 @@
     public static final boolean DEFAULT_SIMUL_SHUNT = false;
     public static final boolean DEFAULT_READ_SLACK_BUS = false;
     public static final boolean DEFAULT_WRITE_SLACK_BUS = false;
+    public static final boolean DEFAULT_VOLTAGE_REMOTE_CONTROLE = false;
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:62]: DEFAULT_VOLTAGE_REMOTE_CONTROL

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 2]
```diff
@@ -100,6 +103,8 @@ protected static void load(LoadFlowParameters parameters, PlatformConfig platfor
                 parameters.setSimulShunt(config.getBooleanProperty("simulShunt", DEFAULT_SIMUL_SHUNT));
                 parameters.setReadSlackBus(config.getBooleanProperty("readSlackBus", DEFAULT_READ_SLACK_BUS));
                 parameters.setWriteSlackBus(config.getBooleanProperty("writeSlackBus", DEFAULT_WRITE_SLACK_BUS));
+                parameters.setVoltageRemoteControl(config.getBooleanProperty("voltageRemoteControl", DEFAULT_VOLTAGE_REMOTE_CONTROLE));
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:106]: DEFAULT_VOLTAGE_REMOTE_CONTROL

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java [hunk 3]
```diff
@@ -103,6 +103,18 @@ public LoadFlowParameters deserialize(JsonParser parser, DeserializationContext
                     parameters.setWriteSlackBus(parser.readValueAs(Boolean.class));
                     break;
 
+                case "voltageRemoteControl":
+                    JsonUtil.assertGreaterThanReferenceVersion(CONTEXT_NAME, "Tag: voltageRemoteControl", version, "1.3");
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java:107]: 1.4?

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java [hunk 4]
```diff
@@ -103,6 +103,18 @@ public LoadFlowParameters deserialize(JsonParser parser, DeserializationContext
...
+                    parser.nextToken();
+                    parameters.setVoltageRemoteControl(parser.readValueAs(Boolean.class));
+                    break;
+
+                case "dc":
+                    JsonUtil.assertGreaterThanReferenceVersion(CONTEXT_NAME, "Tag: dc", version, "1.3");
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java:113]: 1.4?

**Review by annetill — CHANGES_REQUESTED** [review]
Just some small changes.

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java (hunk 3 above)
Comment (tadam50) [reply loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java:107]: 1.3 is right.
Check version is strickly superior to 1.3

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java (hunk 4 above)
Comment (tadam50) [reply loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/json/LoadFlowParametersDeserializer.java:113]: 1.3 is right.
Check version is strickly superior to 1.3

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java (hunk 1 above)
Comment (tadam50) [reply loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java]: Renamed

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java (hunk 2 above)
Comment (tadam50) [reply loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java]: Renamed

**Comment by annetill** [comment]
Can you also treat the ````distributedSlack```` parameter (boolean) and the ````BalanceType```` (enum) ? Thanks!

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 5]
```diff
@@ -58,6 +66,10 @@
     public static final boolean DEFAULT_SIMUL_SHUNT = false;
     public static final boolean DEFAULT_READ_SLACK_BUS = false;
     public static final boolean DEFAULT_WRITE_SLACK_BUS = false;
+    public static final boolean DEFAULT_VOLTAGE_REMOTE_CONTROL = false;
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:69]: I wonder if the default value for voltage remote control could be true. @geofjamg what is your opinion about that ? Do you understand here also remote control for tap changers and swicthes ? Or just for generators and static var compensators ?

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 6]
```diff
@@ -58,6 +66,10 @@
...
+    public static final boolean DEFAULT_DC = false;
+    public static final boolean DEFAULT_DISTRIBUTED_SLACK = false;
```
Comment (annetill) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:71]: I think the default value of distribution slack should be true here. @geofjamg ?

**Review by annetill — COMMENTED** [review]
Just some questions for @geofjamg !

**Review by geofjamg — APPROVED** [review]

**Review by annetill — APPROVED** [review]

**Pull request merged**